src/pytesmo/validation_framework/data_averaging.py

```diff
@@ -0,0 +1,428 @@
+            # iterate from the side of the non-reference
+            for gpi, lat, lon in zip(other_points[0], other_points[1], other_points[2]):
+                # list all non-ref points under the same ref gpi
+                if hasattr(self.ref_class, "grid"):
+                    ref_gpi = self.ref_class.grid.find_nearest_gpi(lon, lat)[0]
+                else:
+                    ref_gpi = self.ref_class.cls.cls.grid.find_nearest_gpi(lon, lat)[0]
+                if ref_gpi in other_lut.keys():
+                    other_lut[ref_gpi].append((gpi, lon, lat))
+                else:
+                    other_lut[ref_gpi] = [(gpi, lon, lat)]
+            # add to dictionary even when empty
+            lut[other_name] = other_lut
+
+        return lut
+
+    def get_timeseries(
+            self,
+            points,
+            other_name,
+    ) -> list:
+        """
+        Get the timeseries for given points info and return them in a list.
+
+        Parameters
+        ----------
+        points: list of tuples
+            list of tuples of (gpi, lon, lat)
+        other_name: str
+            Name of the dataset which the points belong to
+
+        Returns
+        -------
+        dss: list
+            list of dataframes of the reference timeseries
+        """
+        dss = []
+        for gpi, lon, lat in points:
+            # todo: check attributes in loop
+            ds = self.read_ds(other_name, gpi)
+            dss.append(ds)
+
+        return dss
+
+    @staticmethod
+    def temporal_match(
+            to_match,
+            hours=6,
+            drop_missing=False,
+            **kwargs
+    ) -> pd.DataFrame:
+        """
+        Temporal match to the longest timeseries
+
+        Parameters
+        ----------
+        to_match: list
+            list of dataframes to match
+        hours: int
+            window to perform the temporal matching
+        drop_missing: bool, optional. Default is False.
+            If true, only time steps when all points have measurements are kept
+
+        Returns
+        -------
+        matched: pd.DataFrame
+            dataframe with temporally matched timeseries
+        """
+        # get time series with most points
+        ref = to_match[0]
+        for n, df in enumerate(to_match):
+            if df is None:
+                continue
+            points = int(df.count())
+            if int(ref.count()) >= points:
+                continue
+            else:
+                ref = df
+
+        matched = []
+        for series in to_match:
+            matched_ser = temporal_collocation(
+                ref,
+                series,
+                pd.Timedelta(hours, "H"),
+                dropna=True,
+                checkna=True,
+            )
+            matched.append(matched_ser)
+        matched = pd.concat(matched, axis=1)
+
+        if drop_missing:
+            matched.dropna(axis=0, how="any", inplace=True)
+
+        return matched
```

Comment:
There's also `combined_temporal_collocation`, which should probably already do what you want (and a bit more).

https://github.com/TUW-GEO/pytesmo/blob/5e4a54662b935402019b0c232eaae7b79b284494/src/pytesmo/temporal_matching.py#L400